@xen-orchestra/lite/src/composables/host-patches.composable.ts

```diff
@@ -0,0 +1,92 @@
+import type { XenApiHost, XenApiPatch } from "@/libs/xen-api";
+import { useXenApiStore } from "@/stores/xen-api.store";
+import {
+  type MaybeComputedRef,
+  type Pausable,
+  useTimeoutPoll,
+  watchArray,
+} from "@vueuse/core";
+import { computed, ref } from "vue";
+
+export const useHostPatches = (hostRefs: MaybeComputedRef<string[]>) => {
+  const xapiStore = useXenApiStore();
+
+  const timeoutPolls = new Map<string, Pausable>();
+
+  const patchesByHost = ref(new Map<XenApiHost["$ref"], Set<XenApiPatch>>());
+
+  const loadedStatus = ref(new Map<XenApiHost["$ref"], boolean>());
+
+  const patches = computed(() => {
+    const records = new Map<string, XenApiPatch & { $hostRefs: string[] }>();
+
+    patchesByHost.value.forEach((patches, hostRef) => {
+      patches.forEach((patch) => {
+        const id = `${patch.name}@${patch.version}`;
+        const record = records.get(id);
+
+        if (record === undefined) {
+          records.set(id, {
+            ...patch,
+            $hostRefs: [hostRef],
+          });
+        } else {
+          record.$hostRefs.push(hostRef);
+        }
+      });
+    });
+
+    return records;
+  });
+
+  const count = computed(() => patches.value.size);
+
+  const fetchHostPatches = async (hostRef: string) => {
+    if (!patchesByHost.value.has(hostRef)) {
+      patchesByHost.value.set(hostRef, new Set<XenApiPatch>());
+    }
+
+    const missingPatches = await xapiStore.getXapi().getMissingPatches(hostRef);
+
+    patchesByHost.value.get(hostRef)!.clear();
+
+    missingPatches.forEach((missingPatch) => {
+      patchesByHost.value.get(hostRef)!.add(missingPatch);
+    });
+
+    loadedStatus.value.set(hostRef, true);
+  };
+
+  const registerHost = (hostRef: string) => {
+    loadedStatus.value.set(hostRef, false);
+
+    if (timeoutPolls.has(hostRef)) {
+      return timeoutPolls.get(hostRef)!.resume();
+    }
+
+    timeoutPolls.set(
+      hostRef,
+      useTimeoutPoll(() => fetchHostPatches(hostRef), 10000, {
```

Comment:
FYI, subscription in xo-web@5 have a `forceRefresh` method. This allows us to set a long delay for refreshing but still refreshing quickly when we know that the result changed.

Reply:
The difference with XO 5 is that we're sending requests directly to the host instead of through xo-server. So maybe we should be a bit more careful about spamming as we don't control the "spam policy" that the host implements. So I guess 10s or even more is ok, especially for a dashboard. And I don't think a force refresh is necessary because you won't be able to install patches through XO Lite.